Kernel/Devices/Audio/IntelHDA/Stream.cpp

```diff
@@ -0,0 +1,246 @@
+{
+    // 4.5.1: Stream Data In Memory
+    if (bit_size > 16)
+        return 4;
+    else if (bit_size > 8)
+        return 2;
+    return 1;
+}
+
+ErrorOr<void> Stream::initialize_buffer()
+{
+    VERIFY(m_sample_rate > 0);
+    VERIFY(m_pcm_bits > 0);
+    VERIFY(m_channel_count > 0);
+
+    // 4.5.1: Stream Data In Memory
+    // NOTE: we ignore the number of blocks per packet since we are only required to have an integer number
+    //       of samples per buffer, and we always have at least one packet per buffer.
+    size_t block_size_in_bytes = container_size_in_bytes(m_pcm_bits) * m_channel_count;
+    size_t number_of_blocks_in_buffer = PAGE_SIZE / block_size_in_bytes;
+    VERIFY(number_of_blocks_in_buffer > 0);
+
+    size_t number_of_blocks_required_for_cyclic_buffer_size = ceil_div(cyclic_buffer_size_in_ms * m_sample_rate, 1'000);
+    size_t number_of_buffers_required_for_cyclic_buffer_size = AK::max(ceil_div(number_of_blocks_required_for_cyclic_buffer_size, number_of_blocks_in_buffer), minimum_number_of_buffers);
+    VERIFY(number_of_buffers_required_for_cyclic_buffer_size > 0 && number_of_buffers_required_for_cyclic_buffer_size <= 256);
+
+    size_t cyclic_buffer_size_in_bytes = number_of_buffers_required_for_cyclic_buffer_size * PAGE_SIZE;
+
+    TRY(m_buffers.with([&](auto& buffers) -> ErrorOr<void> {
+        buffers = TRY(MM.allocate_dma_buffer_pages(cyclic_buffer_size_in_bytes, "IntelHDA Stream Buffers"sv, Memory::Region::Access::ReadWrite));
+
+        // 3.3.38 Input/Output/Bidirectional Stream Descriptor Cyclic Buffer Length
+        m_stream_io_window->write32(StreamRegisterOffset::CyclicBufferLength, buffers->size());
+
+        // 3.3.39: Input/Output/Bidirectional Stream Descriptor Last Valid Index
+        VERIFY(number_of_buffers_required_for_cyclic_buffer_size <= 256);
+        m_stream_io_window->write16(StreamRegisterOffset::LastValidIndex, number_of_buffers_required_for_cyclic_buffer_size - 1);
+
+        // 3.6.2: Buffer Descriptor List
+        m_buffer_descriptor_list = TRY(MM.allocate_dma_buffer_page("IntelHDA Stream BDL"sv, Memory::Region::Access::ReadWrite));
+        auto bdl_physical_address = m_buffer_descriptor_list->physical_page(0)->paddr().get();
+        m_stream_io_window->write32(StreamRegisterOffset::BDLLowerBaseAddress, bdl_physical_address & 0xffffffffu);
+        m_stream_io_window->write32(StreamRegisterOffset::BDLUpperBaseAddress, bdl_physical_address >> 32);
+
+        // 3.6.3: Buffer Descriptor List Entry
+        auto* buffer_descriptor_entry = m_buffer_descriptor_list->vaddr().as_ptr();
+        for (u8 buffer_index = 0; buffer_index < buffers->page_count(); ++buffer_index) {
+            auto* entry = buffer_descriptor_entry + buffer_index * 0x10;
+            *reinterpret_cast<u64*>(entry) = buffers->physical_page(buffer_index)->paddr().get();
+            *reinterpret_cast<u32*>(entry + 8) = PAGE_SIZE;
+            *reinterpret_cast<u32*>(entry + 12) = 0;
```

Comment:
I think these pointers can be bit-cast instead.

Reply:
Changed!